### PR TITLE
[WIP] Initial work for supporting MCP bypass in Pilot

### PR DESCRIPTION
**Pull request by nmittler**
This is a step toward eliminating MCP (as well as the tcp and serialization overhead) within Pilot for istiod.

Refactors the (previously named) MCP controller to be a Handler-based ConfigStore. Currently, this uses the MCP source in galley as the source of events for the Handler, but will later be refactored to get events directly from a k8s source, bypassing MCP (and network and serialization overhead) entirely.

### CODE REVIEW

File: galley/pkg/config/source/mcp/cache.go [hunk 1]
```diff
@@ -174,16 +167,16 @@ func (c *cache) Dispatch(handler event.Handler) {
 	c.handler = event.CombineHandlers(c.handler, handler)
 }
 
-func (c *cache) sync() {
-	if c.synced || !c.started || !c.fullUpdateReceived {
-		return
-	}
-	c.synced = true
+func (c *cache) sync(resources []*resource.Instance) {
```
Comment (ayj) [inline galley/pkg/config/source/mcp/cache.go]: The previous code exists if the cache is already sync'd. The new code performs a full reset sequence for any change to a collection of resources. Is the intent to emulate the resource batching behavior we had before with snapshots?

**Comment by rshriram** [comment]
to clarify: we need mcp client and server. and mcp client needs to respect the meshConfig.mcpsources settings.

File: pilot/pkg/serviceregistry/mcp/controller.go [hunk 2]
```diff
@@ -125,116 +126,203 @@ func (c *controller) List(typ resource.GroupVersionKind, namespace string) (out
...
-		return fmt.Errorf("apply type not supported %s", change.Collection)
+func (c *controller) Handle(e event.Event) {
+	q := c.syncingMap[e.Source.Name()]
+	if q == nil {
+		log.Warnf("received event for unsupported collection: %s", e.Source.Name())
+		return
+	}
+
+	switch e.Kind {
+	case event.Reset:
```
Comment (howardjohn) [inline pilot/pkg/serviceregistry/mcp/controller.go]: if we get a Reset then FullSync isn't there a period of time that we have no config which can be really bad? I am probably misunderstanding this

Comment (nmittler) [reply pilot/pkg/serviceregistry/mcp/controller.go]: Only for the current update. Basically what we're doing is caching the events between reset and fullsync. So until we get the fullsync, users of the `ConfigStore` API will see the old data before the reset.

File: pilot/pkg/config/event/store.go [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+			log.Warnf(ledgerLogf, err)
+
+		}
+	}
+
+	gvk := schema.Resource().GroupVersionKind()
+	if gvk == serviceEntryKind {
+		s.dispatchServiceEntry(prev, conf, convertEventKind(eventKind))
+	} else {
+		s.xdsUpdater.ConfigUpdate(&model.PushRequest{
```
Comment (Nino-K) [inline pilot/pkg/config/event/store.go]: The old MCP controller always ignore incremental MCP configs (always sent full update), should this event store push updates according to the incremental flag on the config?

**Pull request closed without merging**